code/edgemedia/src/main/java/com/adobe/marketing/mobile/edge/media/internal/MediaEventProcessor.kt

```diff
@@ -11,18 +11,117 @@
+import androidx.annotation.VisibleForTesting
+import com.adobe.marketing.mobile.Event
+import com.adobe.marketing.mobile.edge.media.internal.MediaInternalConstants.LOG_TAG
+import com.adobe.marketing.mobile.edge.media.internal.xdm.XDMMediaEvent
+import com.adobe.marketing.mobile.services.Log
+import java.util.UUID
 
+internal class MediaEventProcessor(
+    private val mediaState: MediaState,
+    private val dispatcher: (event: Event) -> Unit
+) {
+    private val sourceTag = "MediaEventProcessor"
+
+    @VisibleForTesting
+    internal val mediaSessions: MutableMap<String, MediaSession> = mutableMapOf()
+
+    /**
+     * Creates new [MediaSession], assigning it a new random ID.
+     * @return the session ID of the new [MediaSession]
+     */
+    fun createSession(): String {
+        val sessionId = UUID.randomUUID().toString()
+        val session = MediaRealTimeSession(sessionId, mediaState, dispatcher)
+        mediaSessions[sessionId] = session
+        Log.trace(LOG_TAG, sourceTag, "Created new session ($sessionId)")
+        return sessionId
+    }
+
+    /**
+     * Ends the [MediaSession] with the given `sessionId`.
+     * @param sessionId the ID of the [MediaSession] to end
+     * @see [MediaSession.end]
+     */
+    fun endSession(sessionId: String) {
+        val session = mediaSessions[sessionId]
+        if (session != null) {
+            session.end {
+                mediaSessions.remove(sessionId)
+                Log.trace(LOG_TAG, sourceTag, "Successfully ended media session ($sessionId)")
+            }
+        } else {
+            Log.trace(LOG_TAG, sourceTag, "Cannot end media session as session ID ($sessionId) is invalid.")
+        }
+    }
+
+    /**
+     * Queues the [XDMMediaEvent] for processing in the [MediaSession] with ID `sessionId`.
+     * @param sessionId the ID of the [MediaSession] to queue the event
+     * @param event the [XDMMediaEvent] to process
+     * @see [MediaSession.queue]
+     */
+    fun processEvent(sessionId: String, event: XDMMediaEvent) {
+        val session = mediaSessions[sessionId]
+        if (session != null) {
+            session.queue(event)
+            Log.trace(LOG_TAG, sourceTag, "Successfully queued event (${event.xdmData.eventType}) for session ($sessionId)")
+        } else {
+            Log.trace(LOG_TAG, sourceTag, "Cannot queue event (${event.xdmData.eventType}) as session ID ($sessionId) is invalid.")
+        }
+    }
+
+    /**
+     * Sets the `backendSessionId` to the [MediaSession] with matching `requestEventId`.
+     * @see [MediaSession.handleSessionUpdate]
+     */
     fun notifyBackendSessionId(requestEventId: String, backendSessionId: String?) {
+        mediaSessions.forEach { (_, session) ->
+            session.handleSessionUpdate(requestEventId, backendSessionId)
+        }
+
+        // Session may be aborted if backend session ID is invalid
+        mediaSessions.values.removeAll { mediaSession -> !mediaSession.isSessionActive }
```

Comment:
I think it is better to have one way clearing sessions, either in this fashion or we pass handlers to all the methods which can end or abort the session.

Reply:
Kotlin throws an exception when modifying a collection from its `forEach` block, so I cannot remove a session from a `forEach` block. As I wouldn't expect there to be a large number of sessions, I feel the approach of removing inactive sessions in another loop won't impact performance and it keeps the code clean. Another option, however, is using an Iterator instead of the `forEach` function as a MutableIterator does allow for removing elements of the collection.